File: modules/SSHFleet_Py/src/output/statistics.py
# -*- coding: utf-8 -*-
# SSHFleet 统计结果信息模块

import argparse
from typing import Dict, List
from collections import Counter
from datetime import datetime

from src.common.error_handler import error_and_exit_handling_decorator
from src.log import tlog
from src.common.constants import SUCCESS_CATEGORY_EXECUTE, SUCCESS_CATEGORY_TRANSPORT
from src.common.format_utils import get_mode, sort_ips


@error_and_exit_handling_decorator(
    "results_statistics", "计算统计结果信息失败", isexit=True
)
def results_statistics(
    results: List,
    node_infos: List,
    args: argparse.Namespace,
    global_start_time: datetime,
    global_stop_time: datetime,
) -> Dict:
    """
    功能：
        计算统计结果信息

    参数：
        results: 结果列表
        node_infos: 节点信息列表
        args: 命令行参数
        global_start_time: 全局开始时间
        global_stop_time: 全局结束时间

    返回：
        包含所有统计结果信息的字典
    """

    # 基本统计
    results_total = len(results)
    nodeinofs_total = len(node_infos)

    # 成功/失败统计
    exit_counts = Counter(d.get("exit_bool") for d in results)
    success_counts = exit_counts.get(True, 0)
    fail_counts = exit_counts.get(False, 0)

    # 总数校验
    verify = "通过" if nodeinofs_total == results_total else "异常"

    # 分类统计
    category_counts = Counter(d.get("result_category") for d in results)

    # 根据模式移除成功分类
    if get_mode(args) in ("upload", "download"):
        success_category = SUCCESS_CATEGORY_TRANSPORT
        category_counts.pop(success_category, None)
    else:
        success_category = SUCCESS_CATEGORY_EXECUTE
        category_counts.pop(success_category, None)

    # 按数量正（倒）序排序（reverse逆转）失败分类
    sorted_fail_categories = sorted(
        category_counts.items(), key=lambda x: x[1], reverse=True
    )

    # 按分类分组并收集IP地址
    category_ip_map = {}
    for result in results:
        category = result.get("result_category", "未知")
        ip = result.get("ip", "未知IP")
        if category not in category_ip_map:
            category_ip_map[category] = []
        category_ip_map[category].append(ip)

    # 分离成功分类IP
    if "success_category" not in locals():
        tlog.error("success_category 未定义，未知执行模式")
        success_category = "未知"
    success_ips = category_ip_map.pop(success_category, [])

    # 对每个分类的IP进行数字排序
    for category in category_ip_map:
        category_ip_map[category] = sort_ips(category_ip_map[category])

    # 对成功IP进行数字排序
    sorted_success_ips = sort_ips(success_ips) if success_ips else []

    # 构建统计结果字典
    statistics = {
        "results_total": results_total,
        "nodeinofs_total": nodeinofs_total,
        "verify": verify,
        "success_counts": success_counts,
        "fail_counts": fail_counts,
        "sorted_fail_categories": sorted_fail_categories,
        "success_category": (
            success_category if "success_category" in locals() else None
        ),
        "success_ips_count": len(success_ips),
        "sorted_success_ips": sorted_success_ips,
        "category_ip_map": category_ip_map,
        "global_start_time": global_start_time,
        "global_stop_time": global_stop_time,
        "global_cost_time": round(
            (global_stop_time - global_start_time).total_seconds(), 2
        ),
    }
    tlog.success("计算统计结果信息成功")

    return statistics

File: modules/SSHFleet_Go/internal/result/statistics.go
// 结果统计（对位旧 statistics.py）：成功/失败计数、总数校验、失败分类排序、
// 按分类收集 IP、成功分类按模式确定、全局耗时。
package result

import (
	"sort"
	"strconv"
	"strings"
	"time"

	"sshfleet/internal/batch"
	"sshfleet/internal/cli"
	"sshfleet/internal/nodelist"
)

// CategoryCount 分类及其数量。
type CategoryCount struct {
	Category string
	Count    int
}

// Stats 统计结果。
type Stats struct {
	ResultsTotal         int
	NodesTotal           int
	Verify               string // 结果数与节点数一致时「通过」，否则「异常」
	SuccessCounts        int
	FailCounts           int
	SortedFailCategories []CategoryCount
	SuccessCategory      string
	SuccessIPsCount      int
	SortedSuccessIPs     []string
	CategoryIPMap        map[string][]string
	GlobalStartTime      time.Time
	GlobalStopTime       time.Time
	GlobalCostTime       float64
}

// ModeOf 由命令行参数确定执行模式（execute / upload / download）。
func ModeOf(a *cli.Args) string {
	switch {
	case a.Upload != "":
		return "upload"
	case a.Download != "":
		return "download"
	default:
		return "execute"
	}
}

// Statistics 计算结果统计信息。
func Statistics(results *batch.Results, nodes *nodelist.Nodes, a *cli.Args, kw *Keywords, start, stop time.Time) *Stats {
	mode := ModeOf(a)
	successCategory := SuccessCategoryExecute
	if mode != "execute" {
		successCategory = SuccessCategoryTransport
	}

	stats := &Stats{
		ResultsTotal:    results.Len(),
		NodesTotal:      nodes.Len(),
		CategoryIPMap:   map[string][]string{},
		SuccessCategory: successCategory,
		GlobalStartTime: start,
		GlobalStopTime:  stop,
		GlobalCostTime:  stop.Sub(start).Seconds(),
	}
	if stats.NodesTotal == stats.ResultsTotal {
		stats.Verify = "通过"
	} else {
		stats.Verify = "异常"
	}

	counts := map[string]int{}
	for _, r := range results.Items {
		exitBool := r.ExitCode != nil && *r.ExitCode == 0
		if exitBool {
			stats.SuccessCounts++
		} else {
			stats.FailCounts++
		}

		category := Classify(Case{
			ExitCode:     r.ExitCode,
			Error:        strOf(r.Error),
			Output:       r.Output,
			AuthFailure:  strOf(r.AuthFailure),
			Mode:         mode,
			SuccessFiles: r.SuccessFiles,
			FailedFiles:  r.FailedFiles,
		}, kw)

		counts[category]++
		stats.CategoryIPMap[category] = append(stats.CategoryIPMap[category], r.IP)
	}

	delete(counts, successCategory)
	for category, count := range counts {
		stats.SortedFailCategories = append(stats.SortedFailCategories, CategoryCount{category, count})
	}
	sort.SliceStable(stats.SortedFailCategories, func(i, j int) bool {
		return stats.SortedFailCategories[i].Count > stats.SortedFailCategories[j].Count
	})

	successIPs := stats.CategoryIPMap[successCategory]
	delete(stats.CategoryIPMap, successCategory)
	stats.SuccessIPsCount = len(successIPs)
	stats.SortedSuccessIPs = SortIPs(successIPs)
	for category, ips := range stats.CategoryIPMap {
		stats.CategoryIPMap[category] = SortIPs(ips)
	}
	return stats
}

// SortIPs IP 数值序排序（对位旧 sort_ips：按各段数值比较，非法段按字符串兜底）。
func SortIPs(ips []string) []string {
	out := append([]string(nil), ips...)
	sort.SliceStable(out, func(i, j int) bool { return ipLess(out[i], out[j]) })
	return out
}

func ipLess(a, b string) bool {
	as, bs := strings.Split(a, "."), strings.Split(b, ".")
	for i := 0; i < len(as) && i < len(bs); i++ {
		ai, aerr := strconv.Atoi(as[i])
		bi, berr := strconv.Atoi(bs[i])
		if aerr == nil && berr == nil {
			if ai != bi {
				return ai < bi
			}
			continue
		}
		if as[i] != bs[i] {
			return as[i] < bs[i]
		}
	}
	return len(as) < len(bs)
}

func strOf(p *string) string {
	if p == nil {
		return ""
	}
	return *p
}
